Add autofill forms when user registered in project already

Ninefiveblade opened the pull request:
Сделано автозаполнение в форме телеграм, если юзер уже участвует в проекте

AntonZelinsky committed:
Merge pull request #167 from Studio-Yandex-Practicum/fix/report_attempt_number

changed report field is_repeated to attempt_number

Ninefiveblade committed:
Merge branch 'develop' of github.com:Studio-Yandex-Practicum/lomaya_baryery_backend into dev_ivan_feature_forms

Ninefiveblade committed:
add jinja2

ADesBiysk commented on src/api/request_models/user.py:
```diff
@@ -76,3 +84,19 @@ class UserDescAscSortRequest(str, enum.Enum):
+class UserWebhookTelegram(BaseModel):
+    """Валидируем и отдаем данные из базы в Query-форму."""
+
+    name: Optional[StrictStr] = Field(min_length=2, max_length=100)
```
Почему все параметры Optional? У нас не предусмотренно отправка отдельных полей, только всей формы. 
Если убрать Optional, то данная модель практически полностью копирует UserCreateRequest, отсутствует только telegram_id.
Поэтому, надо перенести сюда всё что нужно, и уже от этой модели наследовать UserCreateRequest добавляя нужные поля и валидации.

Ninefiveblade replied:
Тут по другому сделать не получится, из данных может придти None, а эндпоинт зависит от приходящей формы, если сделать так, то поля заполнены не будут (если пользователя нет в базе) и страница регистрации не откроется.

Ninefiveblade replied:
В целом мы в принципе получаем уже отвалидированные данные из базы, так что немного отредактировал то, что было.

ADesBiysk commented on src/api/request_models/user.py:
```diff
@@ -76,3 +84,19 @@ class UserDescAscSortRequest(str, enum.Enum):
+    phone_number: Optional[StrictStr]
+
+    @root_validator
+    def validate_date(cls, values) -> dict:
```
Неудачное название. Здесь не производится валидация даты.
Если убрать Optional со всех полей, то эта валидация не нужна.

ADesBiysk commented on src/api/routers/telegram.py:
```diff
@@ -0,0 +1,33 @@
+from src.api.request_models.user import UserWebhookTelegram
+from src.core.settings import BASE_DIR
+
+templates = Jinja2Templates(directory=BASE_DIR / "src" / "static")
```
Формирование пути перенеси в settings. Здесь только используй его.

Ninefiveblade replied:
Ок

ADesBiysk commented on src/bot/handlers.py:
```diff
@@ -48,12 +51,26 @@ async def start(update: Update, context: CallbackContext) -> None:
+    registration_service = await get_user_service_callback(session)
+    user = await registration_service.get_user_by_telegram_id(telegram_id=telegram_user_id)
+    query = None
+    if user is not None:
```
```suggestion
    if not user:
```

Ninefiveblade replied:
Ок

ADesBiysk commented on pyproject.toml:
```diff
@@ -19,6 +19,7 @@ fastapi-restful = "^0.4.3"
 aiohttp = ">=3.8.3"
 aiolimiter = "1.0.0"
 pytz = "^2022.6"
+jinja2 = "^3.1.2"
```
Не забывай добавлять зависимости в requirements.txt, а то на сервере не заработает.

ADesBiysk commented on src/api/routers/telegram.py:
```diff
@@ -0,0 +1,33 @@
+from src.core.settings import BASE_DIR
+
+templates = Jinja2Templates(directory=BASE_DIR / "src" / "static")
+router = APIRouter(prefix="/telegram", tags=["Telegram template forms"])
```
У нас уже есть роутер hello в котором получаем сообщения от телеграмма. Не имеет смысла их плодить. Логичнее будет тот метод перенести сюда. 
@AntonZelinsky Как думаешь?

Ninefiveblade replied:
> У нас уже есть роутер hello в котором получаем сообщения от телеграмма. Не имеет смысла их плодить. Логичнее будет тот метод перенести сюда. @AntonZelinsky Как думаешь?

У нас там рендерится JSON, а нам нужно рендерить шаблон

Ninefiveblade replied:
Или вы в целом в hello хотите перенести роут?

ADesBiysk replied:
Подожди пока. Не делай.

AntonZelinsky replied:
Я так понял, Денис предлагал объединить с роутером `telegram_webhook`, если да, то я целиком согласен.
Главный роут сделать `/telegram`, два дополнительных `/registration_form` и `/webhook`.
Логика ведь их сама просит объединить.

AntonZelinsky replied:
Только обратил внимание, что у нас же есть проверка `if settings.BOT_WEBHOOK_MODE:`, значит, она должна быть в этом файле, а не в роутере.

ADesBiysk commented:
Завёл отдельную задачу: https://www.notion.so/Bot-523afc1fd1434219a637d92ae59ce05b
Пожалуйста, назначь её на себя и сделай ссылку в шапке PR. 
По задаче. Давай реализуем, то что обсуждали на демо (см. описание в задаче). 
Замечание про перенос роутера пока не делай. Всё остальное нужно делать.

Ninefiveblade committed:
fixes

Ninefiveblade committed:
fixes

Ninefiveblade committed:
fixes

ADesBiysk commented on src/bot/handlers.py:
```diff
@@ -46,14 +49,24 @@ async def start(update: Update, context: CallbackContext) -> None:
+    user = await registration_service.get_user_by_telegram_id(telegram_id=telegram_user_id)
+    query = None
+    if user:
+        query = urllib.parse.urlencode(user.__dict__)
```
Неправильное решение с точки зрения предсказуемости. Сейчас в словаре будет дополнительное значение _sa_instance_state. 
Представим, что в будущем в таблицу users будет добавлено новое поле с ключевым словом или другим подобным значением. В данном варианте, никто не подумает, что нужно переписать этот метод.
Явное лучше неявного. 
Предлагаю в модель User добавить метод формирования словаря с нужными полями.

Ninefiveblade replied:
Сделано

ADesBiysk commented on src/bot/handlers.py:
```diff
@@ -46,14 +49,24 @@ async def start(update: Update, context: CallbackContext) -> None:
             KeyboardButton(
                 text="Зарегистрироваться в проекте",
-                web_app=WebAppInfo(url=settings.registration_template_url),
+                web_app=WebAppInfo(url=WEBAPP_URL.format(url=settings.registration_template_url, query=query)),
```
Нет смысла выносить константу, поскольку переиспользовать её, вероятнее всего, не будем, а формат точно не изменится.
```suggestion
web_app=WebAppInfo(url=f"{settings.registration_template_url}?{query}"),
```

Ninefiveblade replied:
Сделано

ADesBiysk commented on src/bot/handlers.py:
```diff
@@ -46,14 +49,24 @@ async def start(update: Update, context: CallbackContext) -> None:
+    session = get_session()
+    registration_service = await get_user_service_callback(session)
+    user = await registration_service.get_user_by_telegram_id(telegram_id=telegram_user_id)
+    query = None
+    if user:
+        query = urllib.parse.urlencode(user.__dict__)
```
```suggestion
    query = urllib.parse.urlencode(...) if user else None
```

Ninefiveblade replied:
Сделано

ADesBiysk commented on src/api/request_models/user.py:
```diff
@@ -76,3 +84,36 @@ class UserDescAscSortRequest(str, enum.Enum):
+        return value.strftime(DATE_FORMAT)
+
+    @root_validator
+    def validate_fields(cls, values) -> dict:
```
Предлагаю убрать эту проверку. Шаблонизатор не должен падать, если какие-то значения отсутствуют в контексте.
Лучше добавить проверку на наличие значения в самом шаблоне.

Ninefiveblade replied:
Убрано

Ninefiveblade committed:
review_fixes

Ninefiveblade committed:
review_fixes

Ninefiveblade committed:
review_fixes

Ninefiveblade committed:
fixes

Ninefiveblade committed:
fixes

ADesBiysk commented on src/static/registration.html:
```diff
@@ -119,28 +119,28 @@
     <form class="form row">
       <div class="input-field col s12">
-        <input id="name" name="name" type="text" class="validate" minlength="2" maxlength="50" required>
+        <input id="name" name="name" type="text" class="validate" minlength="2" maxlength="50" value="{{ name }}" required>
```
```suggestion
        <input id="name" name="name" type="text" class="validate" minlength="2" maxlength="50" {% if name %} value="{{ name }}" {% endif %} required>
```

Ninefiveblade replied:
Я делаю ее, просто думал не пихать все в 1 мерж

Ninefiveblade replied:
И тут я решил по другому, чтобы не делать кучу проверок шаблоне, заранее проверяю перед отправкой в шаблон

ADesBiysk reviewed (commented):
Не стал делать изменение кнопки и добавление надписи, как обсуждали на демо? Я добавлял описание изменений в задачу.

Ninefiveblade committed:
Обновление регистрационных данных

AntonZelinsky commented on src/bot/handlers.py:
```diff
@@ -31,6 +32,8 @@
 from src.core.services.user_service import UserService
 from src.core.settings import settings
 
+WEBAPP_URL = "{url}?{query}"
```
нигде не используется

AntonZelinsky commented on src/static/registration.html:
```diff
@@ -359,10 +359,17 @@
     };
 
     // send data to server
+    {% if name %}
```
Если Jinja используется только ради этого, то у меня большие сомнения насчёт рациональности её использования. Это легко на js можно обработать. Если не знаешь как, то спроси у кого-нибудь из команды фронта.

AntonZelinsky replied:
Вижу, ещё так `value="{{ phone_number }}"` заполняются данные, это же джинжа? хотя на самом деле, выглядит так, что потом нигде не используется это заполнение данными. Вообще, довольно странно, если оно работает. 
Тем не менее из кода функции отправляющей форму я вижу, что в query параметры передаются все  значения, значит, что они могу быть подставлены средствами js. Предлагаю так и сделать.

И всё равно, я не понимаю как и где jinja работает.

AntonZelinsky commented on src/core/settings.py:
```diff
@@ -63,8 +64,8 @@ def user_reports_dir(self):
-        """Получить ссылку на HTML шаблон регистрации."""
-        return f"{self.APPLICATION_URL}/static/registration.html"
+        """Получить ссылку для на HTML шаблон регистрации."""
+        return f"{self.APPLICATION_URL}/telegram/register_form"
```
По-хорошему, всё-таки нужно с фронтом синхронизоваться, чтобы не получилось так, что они этот роут под себя зарезервируют и будут конфликты.

Hrushon replied:
Не понял, если честно, что нужно сделать

AntonZelinsky commented on src/core/services/user_service.py:
```diff
@@ -53,6 +53,7 @@ def __init__(
 
     async def register_user(self, new_user_data: UserCreateRequest) -> None:
         """Регистрация пользователя. Отправка запроса на участие в смене."""
+        print(new_user_data)
```
🤬🤬🤬
Серьёзно? а про отладчик не слышали?? 
Пока не научитесь пользоваться отладчиком, нельзя считать себя программистом!

AntonZelinsky commented on src/core/db/models.py:
```diff
@@ -127,6 +127,16 @@ class Status(str, enum.Enum):
 
+    @property
+    def object_to_dict(self):
+        return dict(
```
В модели такого быть не должно!

AntonZelinsky commented on src/bot/handlers.py:
```diff
@@ -67,11 +78,19 @@ async def web_app_data(update: Update, context: ContextTypes.DEFAULT_TYPE) -> No
+            )
+        else:
+            await update.message.reply_text(
+                text="Вы успешно обновили ваши регистрационные данные!",
```
По идее, процесс обновления сопровождается с отправкой заявки на новую смену. Нужно это так же обозначать. 
И в сообщении выше (забыл знак препинания) нужно говорить, что пользователь оставил заявку, а не зарегистрировался.

Эти тексты клиент давал?

ADesBiysk commented on src/api/request_models/user.py:
```diff
@@ -101,3 +101,32 @@ class UserDescAscSortRequest(str, enum.Enum):
+    phone_number: Optional[int]
+
+    @validator("date_of_birth")
+    def fix_date_of_birth(cls, value: Union[date, None]) -> Union[str, None]:
```
Для единообразия лучше использовать Optional, иначе не понятно почему здесь именно Union.

Hrushon replied:
странно, что здесь показывают старые данные, почему изменения сюда не подтянулись..
я изменил этот код, там уже нет ни Optional, ни тем более Union

Commit:
fixed bugs with registration form

Commit:
fix dependencies

Commit:
pull origin and merge conflict

Commit:
fix workflow errors

ADesBiysk commented on src/templates/registration/registration.html:
```diff
@@ -1,427 +1,459 @@
-</body>
-
-</html>
+<!DOCTYPE html>
```
Судя по этому файлу, у тебя не настроен пре-коммит. Настрой, пожалуйста, и отправь этот файл ещё раз.

ADesBiysk replied:
Этот файл нужно исправить. Сейчас у него окончания строк LF. У всех остальных файлов в репозитории - CRLF. Из-за этого не отображаются изменения, а помечается, что весь файл изменён. Это очень мешает при проведении код-ревью. Пре-коммит исправляет окончания строк на нужные, но после того как он запрещает отправить файл на сервер, он их исправляет и нужно ещё раз добавить файл в git stage.

Hrushon replied:
я так делал
может проблема в том, что я переместил этот файл. так как он изначально почему-то находился в папке static...я оттуда его перенес туда где ему место - в templates....
может поэтому помечается что файл изменен..у меня в vsc показывает окончания строчек как crlf в этом файле..я снова прогонял его через пре комит и добавлял в гит стаж, но изменения так и не подтянулись к тому файлу, что был в папке static. я в принципе могу написать здесь, какие я строчки изменил:
- добавлены строчки со 114 по 117 - стили для заголовка формы(который меняется в зависимости есть ли юзер в БД)
- добавлены строчки со 123 по 125 - соответственно сам div заголовка формы
- изменены строчки 127, 132, 137, 143, 148 - удалены параметры value= из блоков input
- добавлены строчки 366-368 - получение значений QueryParams в константу params
- добавлены строчки 370-387 - условие проверки: если есть QueryParams, то заполняем форму и делаем соответствующие заголовок формы и надпись кнопки отправки
- добавлена строчка 389 - присваивает соответствующий текст заголовку формы
- изменена строчка 392 у константы showTgButton - вместо текста использована переменная buttontext
Больше я ничего не изменял

ADesBiysk replied:
Перенос файла не должен влиять. Надо попытаться всё настроить, чтобы всем не приходилось делать лишней работы. 
Посмотри какое значение установлено, выполнив команду: `git config --global core.autocrlf`
Должно быть false. Если нет, то попробуй установить, сделать минимальное изменение в файле (например, добавь пробел) и отправить его заново.

Hrushon replied:
поставил false, изменил html и отправил его
если смотреть во вкладке измененные файлы, вроде теперь показываются изменения в этом файле по сравнению с предыдущим.

ADesBiysk commented on src/bot/handlers.py:
```diff
@@ -73,11 +89,20 @@ async def web_app_data(update: Update, context: ContextTypes.DEFAULT_TYPE) -> No
         user_scheme.telegram_id = update.effective_user.id
```
В try ... except должно быть минимальное количество действий. Лучше сделать несколько таких блоков и обрабатывать те исключения, которые могут в нём возникнуть. 
Тогда таких проверок не понадобится: if isinstance(e, ValidationError):

Hrushon replied:
исправил

ADesBiysk commented on src/bot/handlers.py:
```diff
@@ -73,11 +89,20 @@ async def web_app_data(update: Update, context: ContextTypes.DEFAULT_TYPE) -> No
         session = get_session()
         registration_service = await get_user_service_callback(session)
-        await registration_service.register_user(user_scheme)
+        telegram_user_id = update._effective_user.id
```
Вышке на 3 строчки, уже получаем telegram_id. Его и надо использовать дальше.

Hrushon replied:
исправил

ADesBiysk commented on src/bot/handlers.py:
```diff
@@ -52,14 +54,28 @@ async def start(update: Update, context: CallbackContext) -> None:
+    user = await registration_service.get_user_by_telegram_id(telegram_id=telegram_user_id)
+    if user:
+        text = "Подать заявку на участие в смене"
+        web_huck = UserWebhookTelegram.from_orm(user)
```
hook?

Hrushon replied:
исправил

Commit:
Merge branch 'develop' of github.com:Studio-Yandex-Practicum/lomaya_baryery_backend into dev_ivan_feature_forms

Commit:
fix(api, bot, templates): elimination of comments PR

Commit:
fix(bot): litle change order of messages to output

ADesBiysk commented on src/api/routers/telegram.py:
```diff
@@ -0,0 +1,37 @@
+    """
+    Открывает для чтения html-шаблон, в данном случае формы регистрации пользователя.
+
+    Вовзвращает генератор для последующего рендеринга шаблона StreamingResponse-ом.
```
ВоВзвращает

ADesBiysk commented on src/bot/handlers.py:
```diff
@@ -52,14 +53,28 @@ async def start(update: Update, context: CallbackContext) -> None:
+    context: CallbackContext,
+) -> None:
     """Инициализация формы регистрации."""
+    telegram_user_id = update._effective_user.id
```
Здесь точно нет ошибки? Почему обращаемся к защищённому атрибуту, если есть просто effective_user?

Hrushon replied:
халатность, исправил

ADesBiysk commented on src/bot/handlers.py:
```diff
@@ -70,18 +85,25 @@ async def web_app_data(update: Update, context: ContextTypes.DEFAULT_TYPE) -> No
+    except ValidationError as e:
+        e = "\n".join(tuple(error.get("msg", "Проверьте правильность заполнения данных.") for error in e.errors()))
+        await update.message.reply_text(f"Ошибка при заполнении данных:\n{e}")
+    except ValueError as e:
```
В каком случае будет вызвано это исключение? 
В документации используется только ValidationError: https://docs.pydantic.dev/usage/models/#error-handling

Hrushon replied:
действительно, лишний except

ADesBiysk commented on src/bot/handlers.py:
```diff
@@ -70,18 +85,25 @@ async def web_app_data(update: Update, context: ContextTypes.DEFAULT_TYPE) -> No
+        await update.message.reply_text(f"Ошибка при заполнении данных:\n{e}")
+    except ValueError as e:
         await update.message.reply_text(f"Ошибка при заполнении данных:\n{e}")
+    user_scheme.telegram_id = update.effective_user.id
```
Если валидация вызовет исключение, то этого объекта не будет.

Hrushon replied:
исправил

ADesBiysk commented on src/bot/handlers.py:
```diff
@@ -70,18 +85,25 @@ async def web_app_data(update: Update, context: ContextTypes.DEFAULT_TYPE) -> No
+    session = get_session()
+    registration_service = await get_user_service_callback(session)
+    user = await registration_service.get_user_by_telegram_id(telegram_id=user_scheme.telegram_id)
+    if user is None:
+        text = "Процесс регистрации занимает некоторое время - вам придет уведомление."
+    else:
+        text = " Обновленные данные приняты!\nПроцесс регистрации занимает некоторое время " "- вам придет уведомление."
```
```suggestion
    if user:
        text = " Обновленные данные приняты!\nПроцесс регистрации занимает некоторое время - вам придет уведомление."
    else:
        text = "Процесс регистрации занимает некоторое время - вам придет уведомление."
```

Hrushon replied:
действительно...(

Hrushon committed:
Update src/bot/handlers.py

Co-authored-by: DenisAbramenko <adesbiysk@yandex.ru>

Commit:
fix(api, bot, templates): fix comments PR

ADesBiysk commented on poetry.lock:
```diff
@@ -320,41 +320,6 @@ files = [
 tests = ["pytest (>=3.2.1,!=3.3.0)"]
 typecheck = ["mypy"]
```
Этот файл действительно нужно изменять?

Hrushon replied:
по идее нет, я там удалял jinja2, которая была ранее установлена
посмотрю какой командой убрать этот файл из PR и уберу

Hrushon replied:
убрал

ADesBiysk commented on src/bot/handlers.py:
```diff
@@ -70,20 +85,28 @@ async def web_app_data(update: Update, context: ContextTypes.DEFAULT_TYPE) -> No
         user_scheme = UserCreateRequest(**user_data)
+    except ValidationError as e:
+        e = "\n".join(tuple(error.get("msg", "Проверьте правильность заполнения данных.") for error in e.errors()))
+        await update.message.reply_text(f"Ошибка при заполнении данных:\n{e}")
```
Предлагаю сделать проще. После отправки сообщения в except, сделай return, тогда не нужно будет ставить else и сдвигать вправо весь последующий код.

Hrushon replied:
сделал

Ivan-Skvortsov commented on src/bot/handlers.py:
```diff
@@ -52,14 +53,28 @@ async def start(update: Update, context: CallbackContext) -> None:
 
 
-async def register(update: Update, context: CallbackContext) -> None:
+async def register(
```
На мой взгляд, напрашивается рефакторинг этой функции.
Например, как мне кажется, нет смысла получать юзера внутри функции, т.к. мы вызываем её из функции start, в которой [юзер уже используется](https://github.com/Studio-Yandex-Practicum/lomaya_baryery_backend/blob/09bbfdd565b26a8c704292295093082253eb0639/src/bot/handlers.py#L48) и можно его просто передать аргументом.
И можно избавиться от блока else, если перенести тело блока выше if. Но это уже вкусовщина)

Ну и в целом, у нас получается, что функция register может как **регистрировать** пользователя, так и **обновлять** данные уже зарегистрированного пользователя. Это же разный функционал и я бы описал его в разных функцих.

Hrushon replied:
а если объект пользователя не аргументом передавать, а сохранять во временное хранилище context.user_data и далее при удобном случае доставать его, либо None получать?

Hrushon replied:
сделал две функции

Ivan-Skvortsov commented on src/bot/handlers.py:
```diff
@@ -70,20 +85,28 @@ async def web_app_data(update: Update, context: ContextTypes.DEFAULT_TYPE) -> No
         session = get_session()
         registration_service = await get_user_service_callback(session)
-        await registration_service.register_user(user_scheme)
+        user = await registration_service.get_user_by_telegram_id(telegram_id=user_scheme.telegram_id)
```
Если я правильно понял, тут мы дергаем БД только для того, чтобы поменять текстовку сообщения. Действительно ли это нам нужно? Может просто придумать что-то универсальное? Например "Данные приняты!\nПроцесс регистрации занимает некоторое время - вам придет уведомление."

ADesBiysk replied:
Здесь хотелось бы разделить 2 ситуации: регистрация и подача заявки на участие в новой смене.
Рассмотрим ситуацию. Пользователь зарегистрировался и участвовал в одной смене. Потом удалил бота и полгода не участвовал или проект переехал на нового бота. Потом снова услышал про этот проект и решил ещё раз участвовать. Но при регистрации ему отображаются его данные, что может быть неприятным сюрпризом.
Обращение в БД можно убрать, если добавить скрытое поле. Мы же при вызове формы знаем есть пользователь или нет. На логику обработки это поле не повлияет, только на сообщение.
Сообщение лучше немного изменить "Обновленные данные приняты!\nПроцесс обработки заявок занимает некоторое время - вам придет уведомление."
После MVP заказчик изменит текст, если понадобится.

Hrushon replied:
не понял, если честно, про скрытое поле

Hrushon replied:
текст изменил
объект юзера получаю из временного словаря контекста

Ivan-Skvortsov commented on src/bot/handlers.py:
```diff
@@ -70,20 +85,28 @@ async def web_app_data(update: Update, context: ContextTypes.DEFAULT_TYPE) -> No
-        await update.message.reply_text(text=e.detail, reply_markup=ReplyKeyboardRemove())
+        try:
+            await registration_service.register_user(user_scheme)
+        except RegistrationException as e:
```
Получается, что при неудачной регистрации, мы сначала отправим пользователю  сообщение, что все ок (т.е. "Процесс регистрации бла бла..."), а затем сообщение, что в данных есть ошибки? Может быть перенести этот блок выше?

Hrushon replied:
исправил

Hrushon replied:
Не получилось, потому что когда я перенесу его выше, сообщение "Процесс регистрации..." вообще не отправляется, так как кастомные эксепшены перехватывают и отправляют сообщения " Смена начата, вы не успели и прочее...". 
Соответственно, сообщение  о том, что данные получены не появляются при эксепшене совсем и появятся, как я понимаю, только после утверждения, что регистрация пользователя в смену произведена. Соответственно появятся позже, чем это необходимо.

Hrushon replied:
сообщение, что в данных есть ошибки отправляются по большей части выше - в первом блоке try.
здесь, если я правильно понимаю, основную роль играют кастомные исключения

Commit:
choice autocrlf on false

Commit:
remove from PR poetry.lock

Commit:
fix(bot, templates): save user instance in context.user_data

Commit:
fix(bot): fixed bug in docstring

AntonZelinsky commented on src/api/routers/telegram.py:
```diff
@@ -0,0 +1,37 @@
+
+    Возвращает генератор для последующего рендеринга шаблона StreamingResponse-ом.
+    """
+    with open(settings.registration_template_directory.joinpath(FORM_AUTOFILL_TELEGRAM_TEMPLATE), 'rb') as html_form:
```
```suggestion
    with open(settings.registration_template_directory / FORM_AUTOFILL_TELEGRAM_TEMPLATE, 'rb') as html_form:

```

AntonZelinsky replied:
И почему формирование урла разделено на 3 этапа? Сначала приложение находит базовый путь к приложению, потом к папке (это всё в сетингс файле), а уже тут склеивается с именем файла.
Перенеси всё это в сетинги.

AntonZelinsky commented on src/api/routers/telegram.py:
```diff
@@ -0,0 +1,37 @@
+    - **city**: город пользователя
+    - **phone_number**: телефон пользователя
+    """
+    return StreamingResponse(get_register_form(), media_type="text/html")
```
Если метод не используется нигде за пределами этого файла, то должен быть приватным. Или вложенным

Hrushon replied:
исправил

AntonZelinsky commented on src/bot/handlers.py:
```diff
@@ -46,42 +47,74 @@ async def start(update: Update, context: CallbackContext) -> None:
+) -> None:
+    """Инициализация формы для обновления регистрационных данных пользователя."""
+    user = context.user_data.get('user')
+    web_hook = UserWebhookTelegram.from_orm(user)
```
Объект пользователя к тебе приходит из контекста бота, а ты вызываешь метод `from_orm` что-то тут не так. 
И этот метод, я почему-то в исходниках приложения не нашёл.

Hrushon replied:
Этот метод я занял у pydantic, чтобы сериализовать объект пользователя и получить данные для отправки в QueryParams. Объект пользователя был помещен в словарь пользовательских данных контекста в хэндлере `start` после получения его из базы, чтобы не делать лишних запросов к БД в других хэндлерах.

AntonZelinsky commented on src/bot/handlers.py:
```diff
@@ -46,42 +47,74 @@ async def start(update: Update, context: CallbackContext) -> None:
+    update: Update,
+    context: CallbackContext,
+) -> None:
+    """Инициализация формы для обновления регистрационных данных пользователя."""
```
Что-то название функции рассинхронизировано с её описанием и логикой.

Hrushon replied:
не уверен, что правильно, но изменил название

AntonZelinsky commented on src/templates/registration/registration.html:
```diff
@@ -358,12 +363,37 @@
+      document.getElementById("surname").value = surname;
+      var date_of_birth = params.date_of_birth;
+      document.getElementById("date_of_birth").value = date_of_birth;
+      var city = params.city;
```
Выделение не переиспользуемых переменных только усложняет понимание кода, а пользы никакой не несёт.

Hrushon replied:
исправил

AntonZelinsky commented on src/templates/registration/registration.html:
```diff
@@ -358,12 +363,37 @@
+
     const showTgButton = (tgMainButton) => {
-      tgMainButton.setText("Зарегистрироваться в проекте");
+      tgMainButton.setText(buttontext);
```
```suggestion
      tgMainButton.setText(buttonText);
```

AntonZelinsky replied:
js имеет свой стиль именования переменных, нужно его придерживаться

Hrushon replied:
понял

Commit:
fix(api, bot)!: changes routers, registration form

Made changes to registration_form & webhook routers.
Change registration template.
Made changes `settings` about url of webhook & path to registration template.

Commit:
fix(api): change info for documentation endpoint

Made change information for documentation of `webhook` endpoint.

ADesBiysk commented on src/templates/registration/registration.html:
```diff
@@ -1,6 +1,5 @@
-
 <head>
   <meta charset="UTF-8">
   <meta name="viewport" content="width=device-width, initial-scale=1.0">
```
По мотивам проблемы с кэшем. Давай добавим теги для запрета кэширования этого файла.

Hrushon replied:
добавил параметры в headers респонса

ADesBiysk commented on src/bot/handlers.py:
```diff
@@ -47,42 +48,74 @@ async def start(update: Update, context: CallbackContext) -> None:
             KeyboardButton(
                 text="Зарегистрироваться в проекте",
-                web_app=WebAppInfo(url=urljoin(settings.APPLICATION_URL, settings.registration_template_url)),
+                web_app=WebAppInfo(url=f"{settings.registration_template_url}"),
```
settings.registration_template_url возвращает строку. Повторное преобразование не нужно.

ADesBiysk commented on src/core/settings.py:
```diff
@@ -68,8 +67,18 @@ def user_reports_url(self) -> str:
+        return f"{self.APPLICATION_URL}/telegram/webhook"
+
+    @property
+    def registration_template(self):
```
Возвращаем PosixPath.

Hrushon replied:
ой, забыл

ADesBiysk commented on src/bot/handlers.py:
```diff
@@ -47,42 +48,74 @@ async def start(update: Update, context: CallbackContext) -> None:
+        reply_markup=ReplyKeyboardMarkup.from_button(
+            KeyboardButton(
+                text="Подать заявку на участие в смене",
+                web_app=WebAppInfo(url=f"{settings.registration_template_url}?{query}"),
```
Здесь так же.

ADesBiysk commented on src/bot/handlers.py:
```diff
@@ -47,42 +48,74 @@ async def start(update: Update, context: CallbackContext) -> None:
+        text = (
+            " Обновленные данные приняты!\nПроцесс обработки заявок занимает некоторое время - вам придет уведомление."
+        )
+    await update.message.reply_text(
+        text=text,
+        reply_markup=ReplyKeyboardRemove(),
+    )
+    try:
+        await registration_service.register_user(user_scheme)
     except RegistrationException as e:
         await update.message.reply_text(text=e.detail, reply_markup=ReplyKeyboardRemove())
```
Здесь отправку сообщения нужно перенести в самый низ, т.е. когда уже создан новый пользователь, обновлены его данные или найдена ошибка. 
Только нужно пройти по всем методам, которые вызываются в registration_service.register_user и проверить, что все вызываемые исключения наследуются от ApplicationException. Если нет, то надо исправить. Тогда всё будет работать правильно.
```suggestion
    try:
        await registration_service.register_user(user_scheme)
    except ApplicationException as e:
        text=e.detail
    await update.message.reply_text(
        text=text,
        reply_markup=ReplyKeyboardRemove(),
    )

```

Hrushon replied:
Судя по методам, то вызывается только два типа исключения: 
- RegistrationException, наследованное от starlette.HTTPException при добавлении пользователя в смену;
- ValueError при валидации данных пользователя.
В общем, вернул ValueError и RegistrationException

Commit:
fix(api, bot, core): deactivating cache

Deactivating cache in registration`s response.
Refaqtoring the handler `web_app_data`, returned `exception.ValueError`.
Little changes in settings.

Commit:
merging changes in develop

ADesBiysk commented on src/bot/handlers.py:
```diff
@@ -49,44 +50,76 @@ async def start(update: Update, context: CallbackContext) -> None:
-                text="Зарегистрироваться в проекте",
-                web_app=WebAppInfo(url=urljoin(settings.APPLICATION_URL, settings.registration_template_url)),
+                text="Подать заявку на участие в смене",
+                web_app=WebAppInfo(url=settings.registration_template_url + "?" + query),
```
Здесь я не заметил добавление query параметров. Было правильнее. Верни, пожалуйста, f-строку.

ADesBiysk commented on src/bot/handlers.py:
```diff
@@ -49,44 +50,76 @@ async def start(update: Update, context: CallbackContext) -> None:
+        )
+    try:
+        await registration_service.register_user(user_scheme)
+    except (RegistrationException, ValueError) as e:
```
Здесь предлагаю все-таки использовать кастомные исключения.
Сейчас данный метод может вызывать только 2 типа исключений:
1. Наследуется от внешней библиотеки starlette.HTTPException и возвращает HTTP статусы, которые боту не нужны. Т.е. получаем лишнюю зависимость от внешней библиотеки и неиспользуемые атрибуты.
2. Стандартное исключение python, которое, если утрировать, может вызвать любой код проекта. Непонятно, это исключение относится к исключениям вызываемым слоем бизнес логики и мы должны его обработать, как сейчас (отправить сообщение об ошибке пользователю) или ловить выше по уровню все исключения, включая это.

Если метод регистрации в будущем будет усложняться, то нужно будет добавлять сюда новые исключения. Если наследуем все прерывания от одного ApplicationError, то только оно здесь будет.
Только не нужно переопределять стандартное исключение ValueError, нужно заменить его на другое кастомное (добавь своё).

Hrushon replied:
Правильно ли я понял, 
1. Registration Exception необходимо унаследовать от ApplicationException, потому что все равно ApplicationException наследуется от HTTPException;
2. Создать новое исключение, которое будет также наследоваться от ApplicationException и заменит логику ValueError при валидации данных;
3. По итогу здесь перехватывать `ApplicationException` вместо `(RegistrationException, ValueError)`

ADesBiysk replied:
Чтобы не ломать все предыдущие исключения, а сейчас там полный бардак, давай сделаем так: 
0. Добавим базовое исключение ApplicationError, наследуемое от Exception. 
1. Наследуем от ApplicationError.
2. Да. Тоже наследуем от ApplicationError.
3. Перехватываем ApplicationError.

Hrushon replied:
сделал

Commit:
Merge branch 'develop' of github.com:Studio-Yandex-Practicum/lomaya_baryery_backend into dev_ivan_feature_forms

Commit:
feat(core): create `ApplicationError`

Created base exception for app - `ApplicationError`.

Commit:
feat(core): create ApplicationError

Created base error for app - `ApplicationError`.

Ivan-Skvortsov commented on src/api/routers/telegram_webhook.py:
```diff
@@ -1,14 +1,58 @@
+router = APIRouter(prefix="/telegram", tags=["Telegram template forms and webhook"])
+
+
+@router.get(
```
У нас в проекте везде используются cbv, было бы хорошо и здесь придерживаться этого стиля)
Понимаю, что тогда возникнет вопрос с включением/отключением эндпойнта для вебхуков. Мне, если честно, в голову пока приходит только помечать его как deprecated или исключать из схемы, что не повлияет на его работоспособность.

ADesBiysk replied:
Предлагаю смержить в таком виде, а отдельно уже изменить на cbv. Эта задача очень нужна в MVP.
